tee_logger.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+// TeeLogger creates a Logger that duplicates its log calls to two or
+// more loggers. It is similar to io.MultiWriter.
+//
+// For each logging level methad (.Debug, .Info, etc), the TeeLogger calls each
+// sub-logger's level method.
+//
+// Exceptions are made for the Fatal and Panic methods: the returned logger
+// calls .Log(FatalLevel, ...) and .Log(PanicLevel, ...). Only after all
+// sub-loggers have received the message, then the TeeLogger terminates the
+// process (using os.Exit or panic() per usual semantics).
+//
+// DFatal is handled similarly to Fatal and Panic, since it is not actually a
+// level; each sub-logger's DFatal method dynamically chooses to either cal
+// Error or Fatal.
+//
+// Check returns a CheckedMessage chain of any OK CheckedMessages returned by
+// all sub-loggers. The returned message is OK if any of the sub-messages are.
+// An exception is made for FatalLevel and PanicLevel, where a CheckedMessage
+// is returned against the TeeLogger itself. This is so that
+// tlog.Check(PanicLevel, ...).Write(...) is equivalent to tlog.Panic(...)
+// (likewise for FatalLevel).
+func TeeLogger(logs ...Logger) Logger {
```

Comment:
Perhaps just `Tee`? We've got `Sample`, and we'll soon have `Sugar` and `Desugar`, all of which implicitly deal in loggers.

When we get around to teeing outputs, we can use `TeeOutput`.